### DISCUSSION TITLE
Webgrab+

### DISCUSSION THREAD

**Discussion by dfigus** [dfigus/addon-tvheadend#107] [General]
_Originally posted by @obert0 in https://github.com/dfigus/addon-tvheadend/issues/103#issuecomment-1943009144_
>Out of curiosity why did you remove WebGrabPlus from your addon?
>Thank you!

Hi @obert0,

to answer your question from the #103 thread:

I removed Webgrab+ with #77 and I think it had multiple issues:
- It was installed during the initial startup - causing a long runtime
- It was only copying the SiteIniPack files for one country in a hardcoded fashion without a way to configure it
- The SiteIniPack was never updated
- It was installed in the TVHeadend Config folder which bloated the addon-on backups and also the TVHeadend config backup on config migrations

As I also never used it (I just use the internal DVB EPG grabber) , I removed it. If there is a need I would rather add it as part of the Dockerfile so it's backed into the docker image. Are you using it? How do you pull in the correct SiteIniPack for your country? If you help me to understand how you are using it, I'm happy to implement something 😄 .

**Comment by obert0** [comment]
Hi dfigus,
Actually I have setup Tvheadend, wg++ and comskip on a dedicated raspberry pi. I use wg++ because one of my OTA EPG dont work and I can't get recording from this station.

- I dont see issue with runtime because if the folder exist it skip the installation. My only concern is if you have a big guide it take a long time to tvheadend to process it. 
- It is not really hardcoded because it is stored in config file so you can update it easily and it survive rebuild /update. It is also essential to have acces to that folder to edit wg+ config file.
- For a backup perspective I think you only need the active siteini in your config. So you can have only few siteini.

Thank you for your time with this addon!

**Comment by dfigus** [comment]
Hi @obert0,

> * It is not really hardcoded because it is stored in config file so you can update it easily and it survive rebuild /update. It is also essential to have acces to that folder to edit wg+ config file.
> * For a backup perspective I think you only need the active siteini in your config. So you can have only few siteini.

In the original addon the siteini is only copied from India, see code here:
https://github.com/GauthamVarmaK/addon-tvheadend/blob/011abf2aa5b42e9893a2cb085e1dceb7db85d6b1/tvheadend/rootfs/etc/s6-overlay/s6-rc.d/init-tvheadend/run#L28
`cp siteini.pack/India/* siteini.user/`
This is what I meant with "hardcoded". So did you use the siteini from India? Or did you manually copy your proper siteini files? And what about the config you mentioned? Is it similar to the config (ini file) of comskip?

Furthermore I could not spot wg+ in the list of EPG Grabber Modules on TVH, but maybe I overlooked it.

I would see the following a better solution:

- Install wg+ as part of the Dockerfile
- Pull siteini files based on a new addon config option
- Create a new addon config option for the wg+ config file

What else would you see could be improved?

**Reply by obert0** [reply]
Hi @dfigus,

Actually the whole currect siteini is download from WG++:
`&& wget http://webgrabplus.com/sites/default/files/download/ini/SiteIniPack_current.zip \`
I don't really understand why any files should be move to siteini.user folder. I look a little and I don't find the use of this folder. Yes the config file for wg++ is similar to the comskip ini file. It should be adjust to the specific channel you want to grab and must be edit by each user.
The guide here "http://www.webgrabplus.com/documentation/interact-your-pvr-software/tvheadend", is to make the link between TvHeadend and WG++. 
This line:
`&& wget -O /usr/bin/tv_grab_wg++ http://www.webgrabplus.com/sites/default/files/tv_grab_wg.txt \`
add the file that TvHeadend use to map the guide of wg++. Then this line:
`&& sed -i 's|~/.wg++/guide.xml|/config/tvheadend/wg++/guide.xml|g' /usr/bin/tv_grab_wg++ \`
replace the path to point the guide in the "config/tvheanded/wg++" folder.
With those files in place TvHeandend should see the wg++ graber if a guide.xml is accessible.

- Install wg+ as part of the Dockerfile --> I agree!
- Pull siteini files based on a new addon config option --> This cleanup can be done by the user's. siteini are files that should be update sometime.
- Create a new addon config option for the wg+ config file --> This can be a really nice way to do. But the config can be complicated in some cases. See this link as reference: http://www.webgrabplus.com/documentation/configuration/webgrabconfigxml. I think a basic config with only simple channel line like this:
`<channel site="tv-gids.be" site_id="4" update="i" xmltv_id="MP_NatGeo">NATIONAL GEOGRAPHIC</channel>`
can be add in YAML. If more customisation is needed it must be done by editing the file after the installation.

I think this could be a really good start!

**Comment by obert0** [comment]
Hi @dfigus did you still plan to add WG++ on this addon? It can be a really great improvement.

**Reply by dfigus** [reply]
yes, i still plan it, but I did not have the time to do it so far.

**Reply by obert0** [reply]
I understand! Thank you!

**Comment by dfigus** [comment]
@obert0 I implemented a first version in the latest edge build (repo URL https://github.com/dfigus/hassio-addons-edge). Can you have a look at it? I didn't write any docu, yet, but that's on my todo. My first tests with local builds looked good, but as I don't need WG++ I don't really have a good test environment.

**Reply by obert0** [reply]
Hi @dfigus, yes I will be pleased to try it. I manage to  make a local build from your repo to make it work, but I had a hard time with the dependency so I didn't follow your update since then. I will be really glad to merge back to your repo!

**Reply by dfigus** [reply]
I invested around 1 week to find a proper solution. Let me know in case you find something that should be improved there 😀.

**Reply by obert0** [reply]
Hi @dfigus, I install the edge addon. I had no error and I can add the grabber in tvheadend. I couldn't try wg++ because i'm not sure how to make the correct call into the config. Personnally I think the idea of select the country is good to download the right site ini, but the config is a little bit harder to build and need some tweak depending the website you grab. I think a viable way is to add the entire xml line by channel into the config. Or just enter directly into the wg++ config file and update it manually.
I don't understand why, but the wg++ folder is empty at this path: inside addon_config/xxxxxx_tvheadend/tvheadend/wg++. So I cound't test the grabber.

In the build I use, I try a lot to use the last version of WG++, but I end up using V3.3 with mono because I never succed of install Dotnet. I have X86-64 haos version. It is a old version but it work well. Maybe it is less restrictive for some hardware.
I also had issue with automation to regrab the guide. The approch suggest by wg++ is to add a line in the cron table to lauch the grap each day (like you did). So I end up using a "named pipe" inside the addon to lauch some command. It ease a lot the maintnance of wg as it can all be done within ha with automations. It help keep track of the process and take action if an issue occur. 

Thank you very much for your time into this!!! This is a very nice addon. If you need help with the documentation I can help you.

**Reply by dfigus** [reply]
Hi,

you need to add the following config for channel `ard` in site `tvtoday.de` in country `Germany`:

```
wg_channels:
  - country: Germany
    site: tvtoday.de
    channel: ard
```
This will then extract the `ard` channel xml from channels file `Germany/tvtoday.de.channels.xml`. What do you mean with tweaking the website you grab? Can you provide some examples what would be required?

The idea with just providing the entire channel xml sounds good. I could add this as a second expert option in case tweaking is required.

The addon config `wg++` folder will just contain the guide.xml, all of the other parts (incl. config) are are stored in `/wg/.wg++/` as the wg++ installation is already part of the docker build (and does not happen when you install/run the addon on your end). Currently WG++ runs every night at 0:00, in case you want to run it now, just run `/wg/.wg++/run.net.sh`. You could also run it as the startup command for testing😀 .

**Reply by obert0** [reply]
Hi @dfigus, I take a better look and it's not every site that has XML channel. The one I use didn't provide XML. I agree with the expert option for special case.
 Did the config is rebuild at each startup?

I forgot that the wg folder is hide... I will take a look later today! I found the wg at the root of the add-on is that right? 
It can't be run from a console, but not from ha. I should be ok for most use!

**Reply by dfigus** [reply]
> Hi @dfigus, I take a better look and it's not every site that has XML channel. The one I use didn't provide XML. 

My understanding was, that you need to copy the xml entry from the channels file. How would you know how to create the channel entry without the channels.xml file? Can you provide your example?

> Did the config is rebuild at each startup?

The config is written at each startup. All previously manually made changes to that file are reverted.

> I found the wg at the root of the add-on is that right? It can't be run from a console, but not from ha.

It's at `/wg/.wg++/` in the addon. You can't see it outside of the addon, you have to get into the addon (e.g. I always use `docker exec -it addon_xyz_tvheadend bash` via the advanced ssh addon with disabled protection mode).

**Reply by obert0** [reply]
Hi @dfigus ,

This is my config file:

```
<?xml version="1.0"?>
<settings>
   
  <!-- for detailed info about the settings see http://webgrabplus.com/documentation/configuration/webgrabconfigxml 
  and http://webgrabplus.com/sites/default/files/downloads/Misc/Documented_Configuration_Files.zip -->
   
  <filename>/config/tvheadend/wg++/guide.xml</filename>
  <mode></mode>
  <postprocess grab="y" run="n">rex</postprocess>
  <user-agent>Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.130 Safari/537.36 Edg/79.0.309.71</user-agent>
  <!--<decryptkey site="site-name">decrypt_userkey</decryptkey>-->
  <license wg-username="USER" registered-email="EMAIL" password="PASSWORD"></license>
  <logging>on</logging>
  <retry time-out="5">4</retry>
  <timespan>13</timespan>
 
  <!-- 
  GENERATE CHANNEL LIST
   -->
  
  <update>i</update>

  <!-- 
  CURRENT LINEUP
   -->

  <!--
            Replace the next dummy channel entry with the channels you want.
			You can look into the installed siteini.pack folder on your computer
  
            For the latest version,
            see http://webgrabplus.com/epg-channels for the available sites/channels

  -->

	<channel site="tvtv.ca" xmltv_id="CFCF" site_id="CAN-OTAZIPCODE##10104" update="i">CTV</channel>
	<channel site="tvtv.ca" xmltv_id="CFJP-DT" site_id="CAN-OTAZIPCODE##58688" update="i">Noovo</channel>
	<channel site="tvtv.ca" xmltv_id="CIVM-DT" site_id="CAN-OTAZIPCODE##63040" update="i">Télé-Québec</channel>
	<channel site="tvtv.ca" xmltv_id="CKMI" site_id="CAN-OTAZIPCODE##72865" update="i">Global Montréal</channel>
	<channel site="tvtv.ca" xmltv_id="MONTREAL CBFT-DT" site_id="CAN-OTAZIPCODE##45867" update="i">RADIO CANADA</channel>

</settings>
```

I replace my zipcode and my licence credential. I can retrive the site_id by looking at the adress on the website.
I replace the config in the addon and run the guide, but when you change of hardware it need a day to update the hardware id for my wg++ licence. I will try again tomorrow.
It can be really viable for me to add a init command to copy the config from ha folder to the addon for managing complex config that way. Did the init command run before wg_init script?

**Reply by dfigus** [reply]
So a few take aways:

- You also need to be able to configure your license
- Some sites need a decrypt key which need to be configured
- There are other config parameters in the config xml that might need tweaking
Probably it would make sense to even allow supplying a full config XML as the expert option, similar to the `comskip.ini`

Right now the init command is not guaranteed to run after the wg_init script (both depend on base startup but not on each other). I could change that the init commands run after the wg_init.

**Reply by obert0** [reply]
You are right. The config can be customized a lot. The addon config can be limited to what you already done. I think for common case it will be perfect. But there should be a way to push a full config as the expert option. I didn't need custom `comskip.ini`, but it sound like the way to go!

I run the guide and it run well and quite fast. The guide seam complet, but I need to add the link to the channel in TvHeadend to complete the test. I will try as soon as possible, but i'm confident that everything will work at this point.

**Reply by obert0** [reply]
Hi @dfigus, I just set the channels to grap epg from wg++ and it work as expected. Really nice improvement thank you!

**Reply by dfigus** [reply]
Thanks for testing! So I would do the following:

- I will add a new config option (fully optional => hidden by default) to provide a complete config.xml
- Add some documentation

Afterwards I would release a beta version for you so that you can do a final test with your custom config.xml. Do you think it is required to customize the job interval? Right now it's hard-coded to run nightly at 0:00.

**Reply by obert0** [reply]
That's perfect!

For the job interval it's as you which. Maybe it can be useful if a site refresh the guide at another time...

I will also try to add the channel with the config itself.

If you need help for documentation let me know.

**Reply by dfigus** [reply]
I did all of the above and created a v7.0.0b0 beta release in the beta addon repo (https://github.com/dfigus/hassio-addons-beta). Could you give it a try?

**Reply by obert0** [reply]
I install it, but I can only give it a try tomorrow.
To use the complete XML I have to copy/paste the config file into the add-on config?
Thank you!

**Reply by dfigus** [reply]
yes, there is a new config option where you can copy & paste the complete XML. Note: This option is hidden by default and can be shown via the switch to show the additional optional configuration.

**Reply by obert0** [reply]
Hi, I try it really quickly and I got error "403 forbidden" when I try to login even with a fresh install. I will give it another try tomorrow.

**Reply by obert0** [reply]
Ok I succed to log into TVH and load my config manyally. I should wait tomorrow to generate the guide du to new hardware detection... I keep you in touch, but it seam to work. The wg++ config is correctly written inside the folder.

**Reply by obert0** [reply]
Hi @dfigus, I finish the test and it work as expect. I succed to manually run the guide with the config I enter in the addon config. I link the guide to the channel and everything show up as expect.
The only thing I didn't test is the automatic guide regen by the cron file.

**Reply by dfigus** [reply]
@obert0 Thanks for testing this 👏 . Then let's wait for one more day to check if the cron job kicks in. You should also see log entries from the WG++ job. In my test setup I tested this with the job scheduled every 5 min, but would be good if you can double check this as well before releasing.

**Reply by obert0** [reply]
Good idea, I delete the guide and rerun it for a single day. I change it back for a week and I will let it run.

**Reply by obert0** [reply]
Hi @dfigus, The cron job lauch the command as expected.
I face a issue because the .wg++ is rebuilt at each addon startup, but I think it's a problem from my EPG source. When I run the guide a folder robots is created with files from each site that is grabed. Sometime some line in my robots is added with disallow that make the grab fail. This forum talk about this issue: http://www.webgrabplus.com/comment/21586.
This issue is one reason why I use a named pipe to manually remove the disallow line. If theres no robots from my epg provider the robot file is created with the dissalow line and the first time I run wg++ it will fail.
As I sais I think it's not a common problem and I can deal with it. Everything else work perfectly!

**Comment by obert0** [comment]
Hey I have a idea to update a HA sensor when a new recording end. But I will need the homeassistant_api enable in the config. Can you enable that or it has some repercution?
Thank you!

**Reply by dfigus** [reply]
I added the HA API with with the latest v7.0.0b1 release. Can you also have a look at the DOCs? I included a bit of WG++ documentation. If you have further suggestions, please let me know or create a PR 😀 .

**Reply by obert0** [reply]
Thank you very much! The documentation is perfect.
Thank you for your good work this is a solid add-on!

**Reply by dfigus** [reply]
@obert0 I released v7 including WG++ 🎉

**Comment by dfigus** [comment]
@obert0 I just saw the following WG++ errors in my log and I cannot get it fixed:
```
           WebGrab+Plus/w MDB & REX Postprocess -- version  V5.3.0.0            
                                Jan van Straaten                                
                             Francis De Paemeleere                              
 thanks to Paul Weterings, Mat8861, Blackbear199 and all the contributing users 
--------------------------------------------------------------------------------
Job started at 22/01/2025 00:00:01
Checking Internet ..
Job finished at 22/01/2025 00:00:04 done in 3s
Index was outside the bounds of the array.
   at  .‒..ctor(Dictionary`2, Boolean, Boolean, Boolean)
   at  . ..ctor(String)
   at  .—. (String[])
Unhandled Exception
For detailed info, see log file /wg/.wg++/WGLicense.log.txt
Execution stopped
```

As I don't use WG++ I don't have any channels configured (even no dummy one). I added the dummy channel in my config, but again the same error. Also if I use the example config I get the same error. When I added WG++ I did not have this error message... Does it still work on your end?

**Reply by obert0** [reply]
Hi @dfigus,
Since I finish the config with your last release I didn't touch anything and it still run smoothly! (I run it live and it succeed) 
I will look a little bit more to understand your issue.

**Reply by obert0** [reply]
Maybe this can be a issue : http://www.webgrabplus.com/content/https-problem.
It is a old version with mono, but this could be a good start. Did you update your site ini?

**Reply by dfigus** [reply]
@obert0 I did not update the site ini at all, still running the one that was supplied with the latest addon version. Also as I did not add any channels, I wonder whether the site ini files are evaluated at all... I'll keep on digging and maybe add some real channels so check the site ini.

**Reply by obert0** [reply]
@dfigus I didn't study how it works, but was thinking if the site change it will need a update on the site ini. It seem to be a error from what you expect VS what you get didn't match and you overfill a array. Change to another site it will give you a good hint!
Good luck.

**Reply by dfigus** [reply]
@obert0 Ok, I found the issue. The root cause seems the default entry for the license in the `config.xml`:
```
<license wg-username="username" registered-email="forum email-address" password="password"></license>
```
After I added WG++ the behavior of the license check seems to have changed and causes the `Index was outside the bounds of the array.` error. This is also the reason why you don't see any issues with a proper license configured. After I remove the complete license section in the config, the error is not present anymore and the output is without any channels configured:

```

           WebGrab+Plus/w MDB & REX Postprocess -- version  V5.3.0.0            

                                Jan van Straaten                                
                             Francis De Paemeleere                              

 thanks to Paul Weterings, Mat8861, Blackbear199 and all the contributing users 
--------------------------------------------------------------------------------

Job started at 30/01/2025 13:35:54
Checking Internet ..
For License request/update data, see WGLicense.log.txt
Checking License ..
-   Program benefits setting "default"
processing /config/tvheadend/wg++/guide.xml .....

showdetails limitations in effect for license 'default' !!
see WGLicense.log.txt


      i=index  .=same  c=change  g=gab  r=replace  n=new


Group (0) :
update requested for - 0 - out of - 0 - channels for 1 day(s)


Job finished at 30/01/2025 13:35:55 done in 1s
```
I will remove the license from the config, but I can't create a new release due to the current issues with the Ubuntu 24.04 image from Github documented here actions/runner-images#11471

**Reply by obert0** [reply]
Good catch! But it is free of use for most channels. It should work without license.
License give access to some channel and more metadata with the show it grab.
